File: Java/35.SearchInsertPosition.java
class Solution {
	/*
		Link: https://leetcode.com/problems/search-insert-position/
		Problem: 
		Given a sorted array of distinct integers and a target value, return the index if the target is found. If not, return the index where it would be if it were inserted in order.

		Example 1:

		Input: nums = [1,3,5,6], target = 5
		Output: 2
		Example 2:

		Input: nums = [1,3,5,6], target = 2
		Output: 1
		Example 3:

		Input: nums = [1,3,5,6], target = 7
		Output: 4
		Example 4:

		Input: nums = [1,3,5,6], target = 0
		Output: 0
		Example 5:

		Input: nums = [1], target = 0
		Output: 0

		Constraints:
		1 <= nums.length <= 104
		-104 <= nums[i] <= 104
		nums contains distinct values sorted in ascending order.
		-104 <= target <= 104

		Solution:
		Use binary search to find if target in in given sorted array.
		If target is found, return mid.
		After loop exit, check value of low.
		If low is equal to length of the array, it means that target is not found and it is definitely bigger than last element in the array. So, return low.
		If value at index low is lower than target, return low+1
		Else, return low.

		Time Complexity: O(log(n)) where n is size of the array
		Space Complexity: O(1)
	*/
    public int searchInsert(int[] nums, int target) {
        
        int low = 0;
        int high = nums.length;
        int mid = (low+high)/2;
        
        while(low<high)
        {
            mid = (low+high)/2;
            if(nums[mid]==target)
                return mid;
            else if(nums[mid]<target)
            {
                low = mid+1;
            }
            else
            {
                high = mid-1;
            }
        }
        if(low==nums.length)
        {
            return low;
        }
        
        if(nums[low]<target)
        {
            return low+1;
        }
        
        return low;
        
    }
}

File: C++/35. SearchInsertPosition.cpp
// Runtime: 16 ms
// Simple implementation solution

int searchInsert(vector<int>& nums, int target) {
        int n=nums.size();
        if(target<=nums[0])  // It checked if the the number to be inserted is smaller than the first element of the vector and if it is, the number is to be inserted before it. Thus, position which will be 0, to be returned.
            return 0;
        if(target>nums[n-1])  // It checked if the the number to be inserted is larger than the last element of the vector and if it is, the number is to be inserted after it. Thus, position which will be n, to be returned.
            return n;
        for(int i=0;i<n-1;i++)
        {
            if(target>=nums[i] && target<=nums[i+1])  // It checked if the the number lies between any two elements of the vector and if it is, the number is to be placed between them. Thus, position which will be i+1, to be returned.
                return i+1;
        }
        return 0;
    }
